qemu/tests/bridge_vlan.py

```diff
@@ -0,0 +1,262 @@
+                              "server: %s, client: %s" % (server, client),
+                              logging.info)
+        utils_test.run_virt_sub_test(test, params, env, sub_type)
+
+    vms = []
+    sessions = []
+    ifname = []
+    vm_ip = []
+    vm_vlan_ip = []
+    vm_vlan_if = []
+    sub_type = params["sub_type"]
+    host_br = params.get("host_br", "switch")
+    host_vlan_id = params.get("host_vlan_id", "10")
+    host_vlan_ip = params.get("host_vlan_ip", "192.168.10.10")
+    subnet = params.get("subnet", "192.168")
+    mac_str = params.get("mac_str").split(',')
+    os_type = params.get("os_type", "linux")
+
+    linux_modules.load_module("8021q")
+
+    host_vlan_if = "%s.%s" % (host_br, host_vlan_id)
+    if host_vlan_if not in utils_net.get_net_if():
+        host_vlan_if = add_vlan(interface=host_br, v_id=host_vlan_id)
+        if host_vlan_if in utils_net.get_net_if():
+            set_ip_vlan(vlan_if=host_vlan_if, vlan_ip=host_vlan_ip)
+            rm_host_vlan_cmd = params["rm_host_vlan_cmd"] % host_vlan_if
+            funcatexit.register(env, params["type"], _system,
+                                rm_host_vlan_cmd)
+        else:
+            test.cancel("Fail to set up vlan over bridge interface in host!")
+
+    if params.get("start_vm", "yes") == "no":
+        vm_main = env.get_vm(params["main_vm"])
+        vm_main.create(params=params)
+        vm2 = env.get_vm("vm2")
+        vm2.create(params=params)
+        vms.append(vm_main)
+        vms.append(vm2)
+    else:
+        vms.append(env.get_vm([params["main_vm"]]))
+        vms.append(env.get_vm('vm2'))
+
+    for vm_ in vms:
+        vm_.verify_alive()
+
+    for vm_index, vm in enumerate(vms):
+        error_context.context("Prepare test env on %s" % vm.name)
+        session = vm.wait_for_serial_login()
+        if not session:
+            err_msg = "Could not log into guest %s" % vm.name
+            test.error(err_msg)
+
+        if os_type == "linux":
+            interface = utils_net.get_linux_ifname(session,
+                                                   vm.get_mac_address())
+
+            error_context.context("Load 8021q module in guest %s" % vm.name,
+                                  logging.info)
+            session.cmd_output_safe("modprobe 8021q")
+
+            error_context.context("Setup vlan environment in guest %s" % vm.name,
+                                  logging.info)
+            inter_ip = "%s.%s.%d" % (subnet, host_vlan_id, vm_index + 1)
+            set_ip_vlan(interface, inter_ip, session=session)
+            set_arp_ignore(session)
+            error_context.context("Test ping from guest '%s' to host with "
+                                  "interface '%s'" %
+                                  (vm.name, interface), logging.info)
+            try:
+                ping_vlan(vm, dest=host_vlan_ip, vlan_if=interface,
+                          session=session)
+            except NetPingError as details:
```

Comment:
`details` is useless, and again, should we fail the test if ping success, with an `else`

Reply:
@vivianQizhu Thanks, you are right, will update it.